arborist/engine.go

```diff
@@ -0,0 +1,459 @@
+}
+
+// detachRole detaches the role from its parent in the hierarchy and also
+// removes it from the mapping that the engine is tracking.
+func (engine *AuthEngine) detachRole(role *Role) {
+	// Remove the role from its parent's set of subroles. This *assumes* that
+	// the role is not the root role (and therefore must have a parnet), because
+	// the root should never be detached.
+	delete(role.Parent.Subroles, role)
+	// Remove the role from the engine.
+	delete(engine.roles, role.ID)
+}
+
+// detachRole detaches the role from its parent in the hierarchy and also
+// removes it from the mapping that the engine is tracking.
+func (engine *AuthEngine) detachRoleRecursively(role *Role) {
+	engine.detachRole(role)
+	for _, subrole := range role.allSubroles() {
+		delete(engine.roles, subrole.ID)
+	}
+}
+
+// Parameters that constitute an authorization request:
+//
+//     - A list of roles the user possesses
+//     - A list of attempted actions for which the engine must authorize the
+//       user
+//     - A dictionary of constraints which put limits on the attempted action(s)
+//
+// NOTE that the `authRequest.Action.Resource` field will not be initialized by
+// unmarshalling from JSON, because this requires the engine to look up the
+// resource. Parse an `authRequest` using the `AuthEngine.ParseRequest` function.
+type authRequest struct {
+	Roles       []*Role     `json:"roles"`
+	Tags        []string    `json:"tags"`
+	Action      Action      `json:"actions"`
+	Constraints Constraints `json:"constraints"`
+}
+
+func (engine *AuthEngine) ParseRequest(body []byte) (*authRequest, error) {
+	var request *authRequest = &authRequest{}
+	err := json.Unmarshal(body, request)
+
+	// Find the resource for this request.
+	resourceID := request.Action.Resource.ID
+
+	(*request).Action.Resource = engine.findResource(resourceID)
+
+	return request, err
+}
+
+// Struct to contain all the information for an authorization decision issued by
+// the engine.
+type authResponse struct {
+	// Whether or not the request is authorized.
+	Auth bool `json:"auth"`
+
+	// If a role resulted in the granting of authorization, then include its
+	// name in the output.
+	Role_ID *string `json:"role_id"`
+
+	// This field contains the permission that resulted in authorization.
+	PermissionGranting *Permission `json:"permission_matching"`
+
+	// If the request is denied, this field contains a list of permissions
+	// which were relevant to the auth request (that is, had a partly matching
+	// action) but insufficient for authorization. If the request is approved
+	// then this list should be left empty.
+	PermissionsMismatching []*Permission `json:"permissions_mismatching"`
+}
+
+// Process an `authRequest` (which represents a request for authorization on an
+// action, given some roles held by the requester) and return an `authResponse`.
+func (engine *AuthEngine) CheckAuth(request authRequest) authResponse {
+	// Take only the roles with matching tags.
```

Comment:
what's the use case of a service passing tags filter to arborist to evaluate certain action?

Reply:
Hm, fair question. I think perhaps for instance if fence just wants to check auth for group management, then fence would only be interested in roles relevant to `"group"` or something. Though, giving the roles types does handle that instead

Reply:
I don't know if fence ever wants to filter that 🤔  for any auth request, a service wouldn't want to make a decision based on a subset of roles a particular user has? for group management, it's just portal will directly create group by calling arborist and fence will examine if particular group/role exist by calling /role endpoint

Reply:
Ok, if there's no clear use case I'll take it out then